### ISSUE TITLE
Fix GPU tests for repex internal consistency (protein mutations)

### ISSUE THREAD

**Issue by zhang-ivy** [choderalab/perses#1044]
The [repex internal consistency tests](https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py) for terminally blocked amino acid mutations are taking a long time and failing. We need to figure out the minimum number of replicas/iterations necessary to achieve convergence.

I wrote these tests to check 1) for a neutral mutation (ALA->THR), forward and reverse DGs are equal and opposite, 2) for charge changing mutations, the DGs for ARG->ALA->ARG and LYS->ALA->LYS are equal.

Experiment details:
- using the lifting factory
- n_replicas=36
- phase=solvent
- without unsampled endstates
- with LRC on for real-real interactions

Here are the discrepancies (in kT) we are seeing (@ijpulidos helped me run these):

### 500 iterations
replicate 1: -0.637689
replicate 2: -0.18484
replicate 3: 0.631614
### 1000 iterations
replicate 1: 0.254568
replicate 2: 0.210292
replicate 3: -0.0215541

### 2000 iterations
replicate 1: -0.280622
replicate 2: -0.115679
replicate 3: -0.290202
### 3000 iterations
replicate 1: -0.366365
replicate 2: 0.31904
replicate 3: 0.362364
### 4000 iterations
replicate 1: -0.30283
replicate 2: 0.279554
replicate 3: -0.378

Related to: 
https://github.com/choderalab/perses/pull/1031 and https://github.com/choderalab/perses/pull/992

**Comment by zhang-ivy** [comment]
Notes from discussion at yesterday's dev sync:

Define: 
DDG = ALA->THR - THR->ALA
dDDG = sqrt(sum of the squares of ALA->THR and THR->ALA)

Reject if | DDG | > 6 * dDDG # 1 = 68% percentile, 6 is extremely rare
Note this assumes that the stderr is not terribly underestimated

**Comment by zhang-ivy** [comment]
@ijpulidos notes that it takes about 1 hour to run 1000 iterations for ALA->THR (n_replicas=36). We may also want to raise a separate issue about performance concerns. But we can start to investigate the performance issues here a bit. I've asked @ijpulidos to run the same experiments above with the `HybridTopologyFactory` so we can see how much faster it is.

**Comment by ijpulidos** [comment]
Data from neutral mutation is as follows
```
# Neutral mutation RestHTF
iterations DDG 6*dDDG
500 0.689803947734319 1.4975140037656027
1000 0.5941282409570618 1.1851430326984476
2000 0.5905524187826572 0.9496814948346708
3000 0.1664427696842239 0.7198557927254202
4000 0.175336349215101 0.6321557900556232
5000 0.20829390937159786 0.5501208118602601

# Neutral mutation HTF
iterations DDG 6*dDDG
500 0.3490590626887169 1.8280576315676758
1000 0.2176833451410829 1.332873917483699
2000 0.2027259272165125 1.0016163417503057
3000 0.023719950874102835 0.829822104817067
4000 0.3145886147517487 0.7340089087015094
5000 0.12841990003636994 0.6679078946288326
```
It seems that we are nowhere close to the expected accuracy. I wonder if that's even a realistic objective, to aim for a 6-sigma event.

**Comment by zhang-ivy** [comment]
I think it’s actually the opposite. We are just checking to make sure DDG < 6* dDDG, otherwise we reject.

The 6 fold check may be too relaxed actually. We may want to see what happens when we use a 3x check, as that would require the DDG to be within 3 std deviations.

On Jun 15, 2022, at 18:17, Iván Pulido ***@***.***> wrote:

﻿

Data from neutral mutation is as follows

# Neutral mutation RestHTF
iterations DDG 6*dDDG
500 0.689803947734319 1.4975140037656027
1000 0.5941282409570618 1.1851430326984476
2000 0.5905524187826572 0.9496814948346708
3000 0.1664427696842239 0.7198557927254202
4000 0.175336349215101 0.6321557900556232
5000 0.20829390937159786 0.5501208118602601

# Neutral mutation HTF
iterations DDG 6*dDDG
500 0.3490590626887169 1.8280576315676758
1000 0.2176833451410829 1.332873917483699
2000 0.2027259272165125 1.0016163417503057
3000 0.023719950874102835 0.829822104817067
4000 0.3145886147517487 0.7340089087015094
5000 0.12841990003636994 0.6679078946288326


It seems that we are nowhere close to the expected accuracy. I wonder if that's even a realistic objective, to aim for a 6-sigma event.

**Comment by ijpulidos** [comment]
@zhang-ivy ah right, yes, I was understanding this the other way around, thanks for the correction. That said, I don't really understand how is that we are getting "worse" (as in, more "rejectable" results) when we increase the number of iterations. Is that a correct interpretation or am I just missing the point here badly?

**Comment by zhang-ivy** [comment]
@ijpulidos : I've updated the above data to 1) include an additional column with 3 * dDDG and 2) include the charge changing mutation data. 

I was only able to include the data for the `HybridTopologyFactory` and not the `RESTCapableHybridTopologyFactory` because the latter jobs run into memory and wall time issues. 

The `HybridTopologyFactory` data tells me that we are not close to converging -- the difference, 
`ARG->ALA + ALA->ARG -  (LYS->ALA + ALA-> LYS)`, is actually increasing.. I think this means that even though for the neutral mutations, we can get away with 12 replicas, for the charge changing mutations, we need 36 replicas. Can you re-run the charge changing mutations with 36 replicas?

**Comment by ijpulidos** [comment]
@zhang-ivy ok will run them with 36 replicas, also increasing memory and wall time. Thanks!

**Comment by ijpulidos** [comment]
We could try re-running the experiments with 50 steps per iteration and see if that cuts off the wall time to a reasonable amount.

**Comment by zhang-ivy** [comment]
> EDIT: Run time was basically the same for both HTFs. RestHTF: 31619 sec. . Vanilla HTF: 31369 sec.

@ijpulidos : I just realized that these values are for 2 transformations, not 1. That means it actually takes ~ 4 hours per simulation, not 8.

**Comment by ijpulidos** [comment]
@zhang-ivy yes, but the test would still need to perform both transformations, right?

**Comment by zhang-ivy** [comment]
@ijpulidos @jchodera : I've updated [the data above](https://github.com/choderalab/perses/issues/1044#issuecomment-1157006442) with the results from running the neutral test (checking ALA->THR vs THR->ALA) with 50 steps per iteration instead of 250. 

The test, which runs 5000 iterations for ALA->THR and THR->ALA each, takes 24017 sec (6.67 hours) on one lt03 node (GTX 1080Ti). As expected, the DDGs and dDDGs are much larger than they were when we used 250 steps, and it looks like we need the full 5000 iterations to see DDG < 3 * dDDG and 3000 iterations to see DDG < 6 * dDDG. 

Should we just use 50 steps per iteration and n_iterations=3000 for the neutral mutations?

**Comment by ijpulidos** [comment]
> Question: Is it possible to run each leg in parallel on the CI? This would offer a 2x speed up for the neutral test and a 4x speed up for the charge changing test.

It should be possible. We can make `pytest` take care of it, if we create one test for each leg and let `pytest` run the tests in parallel, I think that's achieved with using `pytest-xdist` which I think we are already using.  The problem with that is that it would be using the same gpu for many tests at the same time, so I'm not sure how we want to deal with that.

**Comment by zhang-ivy** [comment]
@jchodera @ijpulidos : I've just updated [the data above](https://github.com/choderalab/perses/issues/1044#issuecomment-1157006442) with the results from the charge changing mutation test (with 50 steps per iteration). 

I think we may be able to get away with 500 or 1000 iterations. If we use 5000 iterations, then the total test time will decrease from 60 hours to 6 hours!
